src/comp/pages/CreateProjectPage.tsx

```diff
@@ -23,90 +38,101 @@ class CreateProjectPage extends React.Component<
     this.state = {
       name: "",
       introduction: "",
-      isNameError: false,
-      isIntroductionError: false,
+      tagIDs: [],
+      sites: [],
+      errors: new Set<ProjectError>([]),
+      isSending: false,
     };
   }
 
-  handleNameInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
-    // multiline属性を消すと表示が乱れるのでこっちで複数行入力を阻止する
-    if (e.target.value.endsWith("\n")) return;
-    const isError = checkBlankSpace(e.target.value) === false;
+  errors = (
+    previousErrors: Set<ProjectError>,
+    error: ProjectError,
+    isError: boolean
+  ) => {
+    const errors: Set<ProjectError> = new Set<ProjectError>([
+      ...Array.from(previousErrors),
+    ]);
+    if (isError) {
+      errors.add(error);
+    } else {
+      errors.delete(error);
+    }
+    return errors;
+  };
+
+  handleNameInputChange = (value: string, isValid: boolean) => {
     this.setState({
-      name: e.target.value,
-      isNameError: isError,
+      name: value,
+      errors: this.errors(this.state.errors, "name", !isValid),
     });
   };
 
-  handleIntroductionInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
-    const isError = checkBlankSpace(e.target.value) === false;
+  handleIntroductionInputChange = (value: string, isValid: boolean) => {
     this.setState({
-      introduction: e.target.value,
-      isIntroductionError: isError,
+      introduction: value,
+      errors: this.errors(this.state.errors, "introduction", !isValid),
     });
   };
 
-  handleSendButton = () => {
-    // TODO: ここでAPIに情報をぶん投げる
-    const projectId = "7438921";
-    this.props.history.push(`/projects/detail/${projectId}`);
+  handleTagChange = (tagIDs: string[]) => {
+    this.setState({
+      tagIDs: tagIDs,
+    });
   };
 
-  render() {
-    let nameHelperText = "";
-    let introductionHelperText = "";
+  handleSitesChange = (sites: PendingSite[]) => {
+    this.setState({
+      sites: sites,
+    });
+  };
+
+  handleSendButton = async () => {
+    this.setState({
+      isSending: true,
+    });
 
-    if (this.state.isNameError) {
-      nameHelperText = "プロジェクト名を入力してください";
+    const projectAPIResult = await ProjectAPI.create(
+      this.state.name,
+      this.state.introduction
+    );
+    const projectId = projectAPIResult.received[0].uuid;
+
+    for (const tagID of this.state.tagIDs) {
+      await TagAPI.belongToProject(tagID, projectId);
     }
-    if (this.state.isIntroductionError) {
-      introductionHelperText = "プロジェクトの説明を入力してください";
+    for (const site of this.state.sites) {
+      await SiteAPI.create(site.description, site.url, projectId);
     }
```

Comment:
```suggestion
    const creationTasks = this.state.sites.map((site) =>
      await SiteAPI.create(site.description, site.url, projectId)
    );
    await Promise.all(creationTasks);
```
これだと作成処理が一つずつ実行されるので、こうしてみたい
(あまり意味がないかもしれない)

Reply:
バックの人と話してみたんですが、`Promise.all`を使って並列でリクエストを送ってしまうとデータベース側でコンフリクトが発生するかもしれないそうです…
ひとまずここは変更せずに残しておきます :bow:

Reply:
👌